### PR TITLE
[RHPAM-2060] - Infinite "Configuration file does not exist" messages …

### PR DESCRIPTION
**Pull request by msmagnanijr**
…in Smart Router logs

See: https://issues.redhat.com/browse/RHPAM-2060

Signed-off-by: Mauricio Magnani <msmagnanijr@gmail.com>

**Thank you for submitting this pull request**

**JIRA**: _(please edit the JIRA link if it exists)_ 

[link](https://www.example.com)

**referenced Pull Requests**: _(please edit the URLs of referenced pullrequests if they exist)_

* paste the link(s) from GitHub here
* link 2
* link 3 etc.

<details>
<summary>
How to retest this PR or trigger a specific build:
</summary>

* <b>a pull request</b> please add comment: <b>Jenkins retest this</b>
 
* <b>a full downstream build</b> please add comment: <b>Jenkins run fdb</b>
  
* <b>a compile downstream build</b> please  add comment: <b>Jenkins run cdb</b>

* <b>a full production downstream build</b> please add comment: <b>Jenkins execute product fdb</b>

* <b>an upstream build</b> please add comment: <b>Jenkins run upstream</b>
</details>

### CODE REVIEW

File: kie-server-parent/kie-server-router/kie-server-router-proxy/src/main/java/org/kie/server/router/repository/ConfigFileWatcher.java [hunk 1]
```diff
@@ -55,6 +56,8 @@ public ConfigFileWatcher(String configFilePath, ConfigurationMarshaller marshall
             if(toWatch.toFile().exists()) {
                 lastUpdate = Files.getLastModifiedTime(toWatch).toMillis();
             } else {
+                File file = new File(this.toWatch.toString() + "kie-server-router.json");
+                file.createNewFile();
```
Comment (spolti) [inline kie-server-parent/kie-server-router/kie-server-router-proxy/src/main/java/org/kie/server/router/repository/ConfigFileWatcher.java:60]: maybe remove this two lines and use only `toWatch.toFile().createNewFile();` instead?

Comment (spolti) [reply kie-server-parent/kie-server-router/kie-server-router-proxy/src/main/java/org/kie/server/router/repository/ConfigFileWatcher.java:60]: plus, can you add a test to make sure the file is created upon object instantiation?

**Review by spolti — CHANGES_REQUESTED** [review]

File: kie-server-parent/kie-server-router/kie-server-router-proxy/src/main/java/org/kie/server/router/repository/ConfigFileWatcher.java (hunk 1 above)
Comment (elguardian) [reply kie-server-parent/kie-server-router/kie-server-router-proxy/src/main/java/org/kie/server/router/repository/ConfigFileWatcher.java:60]: I would create the new file with the configuration provided in the constructor. That would put in sync in memory configuration and file system. 
I would also set the lastUpdate as the new file has.

**Review by elguardian — CHANGES_REQUESTED** [review]
I would create the new file with the configuration provided in the constructor. That would put in sync in memory configuration and file system.
I would also set the lastUpdate as the new file has.

**Comment by elguardian** [comment]
@msmagnanijr is this moving forward ?

**Comment by msmagnanijr** [comment]
> @msmagnanijr is this moving forward ?
Hey @elguardian I'm sorry, last week I had some very high priority tasks and today is a holiday here... so I'll work on it tomorrow. If I have any questions I will let you know.

File: kie-server-parent/kie-server-router/kie-server-router-proxy/src/main/java/org/kie/server/router/repository/ConfigFileWatcher.java [hunk 2]
```diff
@@ -55,10 +58,21 @@ public ConfigFileWatcher(String configFilePath, ConfigurationMarshaller marshall
...
+                try (PrintWriter writer = new PrintWriter(fos)) {
+                    writer.write(cfg);
+                }
+                FileTime lastModified = Files.getLastModifiedTime(toWatch);
+                lastUpdate =  lastModified.toMillis();
                 log.warnv("configuration file does not exist {0} ", this.toWatch);
             }
         } catch (IOException e) {
             log.error("Unable to read last modified date of routers config file", e);
+        } catch (Exception e) {
```
Comment (spolti) [inline kie-server-parent/kie-server-router/kie-server-router-proxy/src/main/java/org/kie/server/router/repository/ConfigFileWatcher.java:74]: Generic exceptions should be final to keep the original stack trace.

File: kie-server-parent/kie-server-router/kie-server-router-proxy/src/main/java/org/kie/server/router/repository/ConfigFileWatcher.java (hunk 1 above)
Comment (spolti) [reply kie-server-parent/kie-server-router/kie-server-router-proxy/src/main/java/org/kie/server/router/repository/ConfigFileWatcher.java:62]: tests?

**Review by spolti — CHANGES_REQUESTED** [review]

**Comment by msmagnanijr** [comment]
Jenkins retest this

File: kie-server-parent/kie-server-router/kie-server-router-proxy/src/main/java/org/kie/server/router/repository/ConfigFileWatcher.java [hunk 3]
```diff
@@ -55,10 +58,23 @@ public ConfigFileWatcher(String configFilePath, ConfigurationMarshaller marshall
...
+                    fos.close();
+                }
+                FileTime lastModified = Files.getLastModifiedTime(toWatch);
+                lastUpdate =  lastModified.toMillis();
                 log.warnv("configuration file does not exist {0} ", this.toWatch);
             }
         } catch (IOException e) {
             log.error("Unable to read last modified date of routers config file", e);
+        } catch (final Exception e) {
+            e.printStackTrace();
```
Comment (elguardian) [inline kie-server-parent/kie-server-router/kie-server-router-proxy/src/main/java/org/kie/server/router/repository/ConfigFileWatcher.java:77]: plz change for log.

File: kie-server-parent/kie-server-router/kie-server-router-proxy/src/main/java/org/kie/server/router/repository/ConfigFileWatcher.java [hunk 4]
```diff
@@ -55,10 +58,23 @@ public ConfigFileWatcher(String configFilePath, ConfigurationMarshaller marshall
             if(toWatch.toFile().exists()) {
                 lastUpdate = Files.getLastModifiedTime(toWatch).toMillis();
             } else {
+                File file = new File(this.toWatch.toString() + "kie-server-router.json");
+                if(file.createNewFile());
```
Comment (elguardian) [inline kie-server-parent/kie-server-router/kie-server-router-proxy/src/main/java/org/kie/server/router/repository/ConfigFileWatcher.java:62]: empty if statement.

File: kie-server-parent/kie-server-router/kie-server-router-proxy/src/main/java/org/kie/server/router/repository/ConfigFileWatcher.java [hunk 5]
```diff
@@ -55,10 +58,23 @@ public ConfigFileWatcher(String configFilePath, ConfigurationMarshaller marshall
...
+                FileOutputStream fos = new FileOutputStream(file);
+                String cfg = marshaller.marshall(configuration);
+                try (PrintWriter writer = new PrintWriter(fos)) {
```
Comment (elguardian) [inline kie-server-parent/kie-server-router/kie-server-router-proxy/src/main/java/org/kie/server/router/repository/ConfigFileWatcher.java:65]: try (FileOutputStream fos = new FileOutputStream(file); PrintWriter writer = new PrintWriter(fos)) 

Then you can remove the close in the finally

**Review by elguardian — CHANGES_REQUESTED** [review]
require some changes.

File: kie-server-parent/kie-server-router/kie-server-router-proxy/src/test/java/org/kie/server/router/repository/FileRepositoryTest.java [hunk 6]
```diff
@@ -213,4 +214,38 @@ public void onConfigurationReloaded() {
...
         repoWithWatcher.clean();
     }
+
+    @Test
+    public void testFileCreation() throws Exception {
+
+        File repositoryDirectory = new File("target" + File.separator + UUID.randomUUID().toString());
+        Configuration configuration = new Configuration();
+        ConfigurationMarshaller marshaller = new ConfigurationMarshaller();
+        File configFile = new File(repositoryDirectory, "kie-server-router.json");
```
Comment (spolti) [inline kie-server-parent/kie-server-router/kie-server-router-proxy/src/test/java/org/kie/server/router/repository/FileRepositoryTest.java:224]: the idea is to test if the file is created if missing.
maybe pass to the constructor only the path without creating the file here?

File: kie-server-parent/kie-server-router/kie-server-router-proxy/src/test/java/org/kie/server/router/repository/FileRepositoryTest.java [hunk 7]
```diff
@@ -213,4 +214,38 @@ public void onConfigurationReloaded() {
...
+                configuration = marshaller.unmarshall(reader);
+
+            } catch (Exception e) {
+                e.printStackTrace();
+            }
+        }
+        
+        ConfigFileWatcher configFileWatcher = new ConfigFileWatcher(configFile.getParentFile().getAbsolutePath(), marshaller, configuration);
+
+        assertNotNull(configFileWatcher);
```
Comment (spolti) [inline kie-server-parent/kie-server-router/kie-server-router-proxy/src/test/java/org/kie/server/router/repository/FileRepositoryTest.java:247]: we need to test if the configuration file is present on fs.

**Review by spolti — CHANGES_REQUESTED** [review]

File: kie-server-parent/kie-server-router/kie-server-router-proxy/src/test/java/org/kie/server/router/repository/FileRepositoryTest.java (hunk 7 above)
Comment (msmagnanijr) [reply kie-server-parent/kie-server-router/kie-server-router-proxy/src/test/java/org/kie/server/router/repository/FileRepositoryTest.java:247]: @spolti I'm not sure how to do this using this constructor but maybe I can validate using this. https://github.com/kiegroup/droolsjbpm-integration/blob/1753048d3d0419b7a06540eafc14a60fa1250a99/kie-server-parent/kie-server-router/kie-server-router-proxy/src/main/java/org/kie/server/router/repository/FileRepository.java#L47  and so doing some check.

File: kie-server-parent/kie-server-router/kie-server-router-proxy/src/test/java/org/kie/server/router/repository/FileRepositoryTest.java (hunk 6 above)
Comment (msmagnanijr) [reply kie-server-parent/kie-server-router/kie-server-router-proxy/src/test/java/org/kie/server/router/repository/FileRepositoryTest.java:224]: Okay, I will try this approach.

**Pull request closed without merging**